synapse/storage/controllers/persist_events.py

```diff
@@ -150,45 +173,46 @@ def __init__(
     async def add_to_queue(
         self,
         room_id: str,
-        events_and_contexts: Iterable[Tuple[EventBase, EventContext]],
-        backfilled: bool,
+        task: _EventPersistQueueTask,
     ) -> _PersistResult:
-        """Add events to the queue, with the given persist_event options.
+        """Add a task to the queue.
 
-        If we are not already processing events in this room, starts off a background
+        If we are not already processing tasks in this room, starts off a background
         process to to so, calling the per_item_callback for each item.
 
         Args:
             room_id (str):
-            events_and_contexts (list[(EventBase, EventContext)]):
-            backfilled (bool):
+            task (_EventPersistQueueTask): A _PersistEventsTask or
+                _UpdateCurrentStateTask to process.
 
         Returns:
             the result returned by the `_per_item_callback` passed to
             `__init__`.
         """
         queue = self._event_persist_queues.setdefault(room_id, deque())
 
-        # if the last item in the queue has the same `backfilled` setting,
-        # we can just add these new events to that item.
-        if queue and queue[-1].backfilled == backfilled:
+        # try to batch `_PersistEventsTask`s with the same `backfilled` setting together.
+        if (
+            queue
+            and isinstance(task, _PersistEventsTask)
+            and isinstance(queue[-1].task, _PersistEventsTask)
+            and queue[-1].task.backfilled == task.backfilled
+        ):
             end_item = queue[-1]
+            existing_task = queue[-1].task
+            # add our events to the existing queue item
+            existing_task.events_and_contexts.extend(task.events_and_contexts)
```

Comment:
I wonder if it'd be cleaner to have a `def try_update(..) -> bool` function in `_EventPersistQueueTask` that encapsulates this logic? For `_UpdateCurrentStateTask` it'd simply always return false (i.e. update failed), and  in `_PersistEventsTask` have this check? Though that might complicate things unnecessarily.

Reply:
I had a go at this